avocado/core/requirements.py

```diff
@@ -0,0 +1,98 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2020
+# Authors: Cleber Rosa <crosa@redhat.com>
+
+"""
+Test requirements module.
+"""
+
+import os
+import sqlite3
+
+from .data_dir import get_datafile_path
+
+#: The location of the requirements cache database
+CACHE_DATABASE_PATH = get_datafile_path('cache', 'requirements.sqlite')
+
+#: The definition of the databse schema
+SCHEMA = [
+    'CREATE TABLE IF NOT EXISTS requirement_type (requirement_type TEXT UNIQUE)',
+    'CREATE TABLE IF NOT EXISTS environment_type (environment_type TEXT UNIQUE)',
+    ('CREATE TABLE IF NOT EXISTS environment ('
+     'environment_type TEXT,'
+     'environment TEXT,'
+     'FOREIGN KEY(environment_type) REFERENCES '
+     'environment_type(environment_type)'
+     ')'),
+    ('CREATE UNIQUE INDEX IF NOT EXISTS '
+     'environment_idx ON environment (environment, environment_type)'),
+    ('CREATE TABLE IF NOT EXISTS requirement ('
+     'environment_type TEXT,'
+     'environment TEXT,'
+     'requirement_type TEXT,'
+     'requirement TEXT,'
+     'FOREIGN KEY(environment_type) REFERENCES environment(environment_type),'
+     'FOREIGN KEY(environment) REFERENCES environment(environment),'
+     'FOREIGN KEY(requirement_type) REFERENCES requirement_type(requirement_type)'
+     ')'),
+    ('CREATE UNIQUE INDEX IF NOT EXISTS requirement_idx ON requirement '
+     '(environment_type, environment, requirement_type, requirement)')
+]
```

Comment:
Based on this code and on this commit message I would choose something that uses more like an "MVC pattern" and also would try to avoid a DB scheme here.

We don't need to discuss the choice for SQLite here, but my suggestion would be to split this into a proper package (for instance `avocado.core.persistence.backends.sqlite`) with a very basic interface, so if we decide to change this later, we could do it easily. Also for better organization, avoiding specific SQLite code inside the `requirements` module.

Reply:
> Based on this code and on this commit message I would choose something that uses more like an "MVC pattern" and also would try to avoid a DB scheme here.
> 
> We don't need to discuss the choice for SQLite here, but my suggestion would be to split this into a proper package (for instance `avocado.core.persistence.backends.sqlite`) with a very basic interface, so if we decide to change this later, we could do it easily. Also for better organization, avoiding specific SQLite code inside the `requirements` module.

You're supposing that we will want to support multiple backends?  IMO there's no proof that we need multiple backends to solve the same problem... it would only increase our maintenance effort at this stage.

Then, we are free to change this interface and implementation at any time: it's pretty much an internal interface, it's "opt-in" (that is, not required that anyone uses) and it's limited to 2 simple functions.

Reply:
Even if we don't have multiple backends, IMO is valid the split for better organization.

Reply:
I think splitting it will also make it more readable as the current code will be abstracted away from such implementation details.

Reply:
So this is what the split *change* would look like:

https://github.com/clebergnu/avocado/commit/34d40f800832012c0009ce372042091d4a2f66aa

And this is what the entirety of the resulting `avocado.core.requirements` would look like (minus comments):

```
from .requirementbackend import (set_requirement_on_cache,
                                 get_requirement_on_cache)
```

I can go any way, but this looks like over engineering at this point, and IMO, makes it more confusing to follow the code.

Do you guys think it's worth it?

Reply:
I am not sure about the merged `requirementbackend` module name (probably best to have it as `requirements_backend.sqlite` or something) but I am still for this pattern. My main motivation is that this is an area of the code that will no doubt grow a lot when the we discuss and establish more of the requirements resolver blueprint and even now we could be fairly certain that the requirements definitions and checking/validation would need more implementations to come.

So indeed, it could look a bit like over-engineering at the moment to have a list with one item of supported backends but (1) it will quite certainly be needed very soon and (2) it is a good pattern by itself and even with one item we can see that your suggested changes were mostly separating the code into another module and leaving the first module easier to read and abstracted from the second.

Reply:
About the naming, I know @beraldoleal doesn't like names with underscores (and he's got a point because Python suggests against it).  So, I'm finding it hard to name it because something like `requirements.backends.sqlite` would imply that we have and select one of the available backends, while we dont...

Suggestions on top of this comment?

Reply:
Good point. The `requirements.backends.sqlite` is clean of underscores so I guess the only trouble with it is that there are no modules other than `sqlite` which is more or less the single item list argument we mentioned above but in terms of files. In this case I would still go for it, and we will gradually populate the `requirements.backends` subpackage with more options.

The module `requirements` itself could just have an attribute for the selected backend:
```python
# inside the requirements.py:

from .backends import *

BACKEND = sqlite

# more code here
```

Forgive me if this example is too simplistic.

Reply:
I can go the `requirements.backends.sqlite` route... but I don't really see more than *one* backend being supported... I can think of a use case right now.  This is what is hurting me.

Maybe, if we switch the `backend` for `cache`, it could simply be, at this time `avocado.core.requirements.cache`, with a single implementation at this time.  *IF* we get a second implementation, we introduce the module `backends` namespace.

How does that sound?

Reply:
It is not as explicit as `sqlite` but `cache` sounds reasonable middle ground considering the scarcity of backends and backend types we have at the moment. I have an idea about a requirements backend to suggest but it is in its early "drafts" and I would rather first focus on suggesting spawners and models of isolation. Some requirements backends might also end up entangled with some specific spawners so I am fine with your current choice before I could say more :-1:

Reply:
> I can go the `requirements.backends.sqlite` route... but I don't really see more than _one_ backend being supported... I can think of a use case right now. This is what is hurting me.

Also what matters is that the `requirements` module is left free of assumptions about cache and other implementations so this should be enough for the time being.

Reply:
> So, I'm finding it hard to name it because something like requirements.backends.sqlite would imply that we have and select one of the available backends, while we don't...

```python
from avocado.core.backends.sqlite import (methods here)
``` 

Probably we are going to disagree here too. :) lol But I don't think it is a problem to have this organization even with only one backend supported. But if you don't like that, maybe we could have:

```python
from avocado.utils.sqlite import (methods here)
```

or 

```python
from avocado.core.sqlite import (methods here)
```

where methods are not related to "requirements", would be something like query, save, create_table, drop, delete, ....

Reply:
Well, the problem is that we currently don't have (and this don't need) utilities for query, save, create_table, drop, delete. Python's dbapi and `sqlite3` module is rather straightforward to use.

Reply:
I'm also on the side that any direct interaction with a database should not be in this module. The methods responsible to interact with the `cache`/backend should be abstract enough to do not break this module in case we change the backend.

Reply:
OK, I've reorganized the module structure on v2.